File: surf-api-bukkit/surf-api-bukkit-plugin-test/src/main/java/dev/slne/surf/surfapi/bukkit/test/BukkitPluginMain.java
package dev.slne.surf.surfapi.bukkit.test;

import dev.jorel.commandapi.CommandAPI;
import dev.slne.surf.surfapi.bukkit.api.inventory.framework.InventoryFrameworkExtensions;
import dev.slne.surf.surfapi.bukkit.api.packet.listener.SurfBukkitPacketListenerApi;
import dev.slne.surf.surfapi.bukkit.test.command.SurfApiTestCommand;
import dev.slne.surf.surfapi.bukkit.test.command.subcommands.inventory.TestInventoryView;
import dev.slne.surf.surfapi.bukkit.test.command.subcommands.reflection.Reflection;
import dev.slne.surf.surfapi.bukkit.test.config.ModernTestConfig;
import dev.slne.surf.surfapi.bukkit.test.listener.ChatListener;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

public class BukkitPluginMain extends JavaPlugin {

  @SuppressWarnings("unused")
  public static @NotNull BukkitPluginMain getInstance() {
    return getPlugin(BukkitPluginMain.class);
  }

  @Override
  public void onLoad() {
    ModernTestConfig.Companion.init();
    ModernTestConfig.Companion.randomise();

    SurfBukkitPacketListenerApi.Companion.getInstance().registerListeners(new ChatListener());
    InventoryFrameworkExtensions.register(TestInventoryView.INSTANCE);
//    TestConfig config = SurfCore#Api.getCore()
//        .createModernYamlConfig(TestConfig.class, getDataFolder().toPath(), "test.yml");
//    TestConfig2 config2 = SurfCoreApi.getCore()
//        .createModernYamlConfig(TestConfig2.class, getDataFolder().toPath(), "test2.yml");
//
//    System.err.println(config.testLinkedList);
//    System.err.println(config.testObjectLinkedList);
  }

  @Override
  public void onEnable() {
    new SurfApiTestCommand().register();
    Reflection.class.getClassLoader(); // initialize Reflection
  }

  @Override
  public void onDisable() {
    CommandAPI.unregister("surfapitest");
  }
}

File: surf-api-bukkit/surf-api-bukkit-plugin-test/src/main/kotlin/dev/slne/surf/surfapi/bukkit/test/BukkitPluginMain.kt
package dev.slne.surf.surfapi.bukkit.test

import com.github.shynixn.mccoroutine.folia.SuspendingJavaPlugin
import dev.jorel.commandapi.CommandAPI
import dev.slne.surf.surfapi.bukkit.api.inventory.framework.register
import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.packet.listener.packetListenerApi
import dev.slne.surf.surfapi.bukkit.test.command.SurfApiTestCommand
import dev.slne.surf.surfapi.bukkit.test.command.subcommands.inventory.TestInventoryView
import dev.slne.surf.surfapi.bukkit.test.command.subcommands.reflection.Reflection
import dev.slne.surf.surfapi.bukkit.test.config.ModernTestConfig
import dev.slne.surf.surfapi.bukkit.test.listener.ChatListener
import dev.slne.surf.surfapi.core.api.component.surfComponentApi

@OptIn(NmsUseWithCaution::class)
class BukkitPluginMain : SuspendingJavaPlugin() {
    override suspend fun onLoadAsync() {
        ModernTestConfig.init()
        ModernTestConfig.randomise()

        surfComponentApi.load(this)
        packetListenerApi.registerListeners(ChatListener())
        TestInventoryView.register()
    }

    override suspend fun onEnableAsync() {
        SurfApiTestCommand().register()
        Reflection::class.java.getClassLoader() // initialize Reflection

        surfComponentApi.enable(this)
    }

    override suspend fun onDisableAsync() {
        CommandAPI.unregister("surfapitest")
        surfComponentApi.disable(this)
    }

    companion object {
        val instance: BukkitPluginMain
            get() = getPlugin(BukkitPluginMain::class.java)
    }
}

val plugin get() = BukkitPluginMain.instance